### ISSUE TITLE
bug:  no validation is present in bindingpolicy CRDs for label selectors

### ISSUE THREAD

**Issue by pdettori** [kubestellar/kubestellar#1840]
### Describe the bug

It is currently possible to create bindingpolicies with invalid labels in label selector. That can be detected only looking at the kubestellar-controller-manager logs:

```
E0301 13:34:57.225776 1 controller.go:424] error matching selectors: values[0][name]: Invalid value: "edgeplatform-test-wec1, edgeplatform-test-wec2": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue', or 'my_value', or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')
```

### Steps To Reproduce

1. create a binding policy as follows:

```
apiVersion: control.kubestellar.io/v1alpha1
kind: Placement
metadata:
  name: nvidia-gpu-placement
spec:
  clusterSelectors:
  - matchLabels:
      name: edgeplatform-test-wec1, edgeplatform-test-wec2
  downsync:
  - objectSelectors:
    - matchLabels:
        app.kubernetes.io/part-of: wec-gpu
  wantSingletonReportedState: true
 ``` 
 
 2. look at kubestellar-controller-manager logs

### Expected Behavior

Should validate and prevent user to create such binding policy.

### Additional Context

A similar experiment with a deployment shows that there is validation for label selectors:

Apply a deployment as follows:

```
piVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx, nginx2
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80

(⎈|kind-kubeflex:N/A)$ k apply -f ~/Desktop/test1.yml 
The Deployment "nginx-deployment" is invalid: 
* spec.selector.matchLabels: Invalid value: "nginx, nginx2": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')
* spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app":"nginx, nginx2"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: invalid label selector
```

**Comment by pdettori** [comment]
We should investigate if we can use the `// +kubebuilder:validation:Pattern string` when generating the CRD (in https://book.kubebuilder.io/reference/markers/crd-validation).

**Comment by ezrasilvera** [comment]
@pdettori Sounds reasonable.  When fixing thsi , let's make sure we have a unit test for "illegal selector  strings"

**Comment by ezrasilvera** [comment]
@waltforme Till you fix it, can you please add a bullet into our know issues section

**Comment by waltforme** [comment]
@ezrasilvera I did some initial investigation. This issue may not be trivial as it appears to be.

From @pdettori 's examples, we see that a Deployment's invalid `.spec.selector` is not accepted, but a BindingPolicy's invalid `.spec.clusterSelectors` is accepted.

The difference is because the apiserver validates a Deployment and a BindingPolicy in different ways. Let me elaborate as follows.

For a Deployment, below is how it is validated.
https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/pkg/registry/apps/deployment/strategy.go#L86-L90

We see that a Deployment is validated thoroughly, for example, the `.spec.selector` is validated 'strictly'.
https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/pkg/apis/apps/validation/validation.go#L558

For a BindingPolicy, below is how it is validated.
https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L204-L230

We see that a BindingPolicy is not validated as thoroughly as a Deployment. Several things are validated, one of them is the validation against the schema.
https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L54

Look at our schema, we see that the schema only says `type: object`, for a `metav1.LabelSelector`'s `MatchLabels`.
https://github.com/kubestellar/kubestellar/blob/1f75c1f7919e8dda2acc2d1135602ce012313018/config/crd/bases/control.kubestellar.io_bindingpolicies.yaml#L94-L102

So here is the problem. Even we explicitly say that `.spec.clusterSelectors` is a slice of `metav1.LabelSelector` here
https://github.com/kubestellar/kubestellar/blob/1f75c1f7919e8dda2acc2d1135602ce012313018/api/control/v1alpha1/types.go#L27,
each `metav1.LabelSelector` is still not validated as strictly as
https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/pkg/apis/apps/validation/validation.go#L559.

In other words, as long as we continue to reuse the `metav1.LabelSelector` struct (surely we want to),
```
	ClusterSelectors []metav1.LabelSelector `json:"clusterSelectors,omitempty"`
```
We will never get the label values of `MatchLabels` strictly validated.

For a solution. I suggest that we encourage our users to use a `metav1.LabelSelector`'s `MatchExpressions` instead of `MatchLabels`, in our documentation.

**Comment by pdettori** [comment]
thank you @waltforme for the analysis. Maybe I am missing something, but could we use https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/apis/apps/validation/validation.go#L559 to validate our selector?

**Comment by MikeSpreitzer** [comment]
@pdettori : I think what @waltforme is saying is that we do not get to write the validation code, that is constant in https://github.com/kubernetes/kubernetes/blob/ebc1a7b7fb52f30021b7e66f77070dac8e6b839b/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L204-L230

**Comment by MikeSpreitzer** [comment]
(A) The slice of label selectors is not a string, it is a slice of structure things. `kubebuilder:validation:Pattern` will not help us here.

(B) I keep forgetting what can go in a use of `kubebuilder:validation:XValidation`. Someone should check, but I expect disappointment.

(C) If kubebuilder supported _and_ (which IS in OpenAPI v3) then we could add constraints beyond what are built into `metav1.LabelSelector`.

(D) If you think that `metav1.LabelSelector` is buggy: (1) are you really sure? How could it be buggy for so long? (2) we can propose a fix (yes, that will not get back to us quickly, but it is ultimately the best answer).

(E) #1950 introduces a better way to report user errors. We want to get #1950 merged, right?

**Comment by MikeSpreitzer** [comment]
https://opensource.googleblog.com/2023/11/kubernetes-crd-validation-using-cel.html gives me a bit of hope. It is not useful until we require Kubernetes release 1.29 or greater. At that point, I think it is trying to say that our CRD can add to the requirements of `metav1.LabelSelector`.

**Comment by MikeSpreitzer** [comment]
(G) we can modify the OpenAPI post-generation, either through the kustomization or something else.

**Comment by MikeSpreitzer** [comment]
Which version of KubeStellar has the line number in the opening comment?

**Comment by MikeSpreitzer** [comment]
Regarding (G): https://json-schema.org/draft/2020-12/json-schema-core#name-keyword-independence says this would be fairly easy, needing only _addition_ of material into the schema.

Regarding (E): actually #2044 is the relevant PR here, not #1950. The next shoe to drop is actually populating the "status" subresource, including the `ObservedGeneration`.

**Comment by waltforme** [comment]
Thanks @MikeSpreitzer for these suggestions around CEL based validations. I started with (G) because it sounds most promising. I have some good news and bad news.

#### Good news
Good news is that we can indeed add extra validations to the generated BindingPolicy schema.

In this example https://github.com/waltforme/kubestellar/commit/540a31193157f0467cf9ddf3d084b53d8718c5e2, I was able to introduce extra validations to the values of the `matchLabels` map.

The extra validations work as below. The `valid` bindingPolicy passes the validation and the `invalid` doesn't pass.
```
ubuntu@ip-172-31-5-54:kubestellar$ kc create -f - <<EOF                                                   apiVersion: control.kubestellar.io/v1alpha1
kind: BindingPolicy
metadata:
  name: valid
spec:
  clusterSelectors:
  - matchLabels: {"valid-key": "start-of-value-1"}
  downsync:
  - objectSelectors:
    - matchLabels: {"app.kubernetes.io/name":"nginx"}
---
apiVersion: control.kubestellar.io/v1alpha1
kind: BindingPolicy
metadata:
  name: invalid
spec:
  clusterSelectors:
  - matchLabels: {"in-valid-key": "begin-of-value-1"}
  downsync:
  - objectSelectors:
    - matchLabels: {"app.kubernetes.io/name":"nginx"}
EOF
bindingpolicy.control.kubestellar.io/valid created
The BindingPolicy "invalid" is invalid: spec.clusterSelectors[0]: Invalid value: "object": KS additional validation on matchLabels, a value must starts with 'start-of-value'
```

#### Bad news
Bad news is that the 'estimated rule cost' of the extra validations in BindingPolicy's schema easily exceeds the cost budget.

Changing the same example https://github.com/waltforme/kubestellar/commit/540a31193157f0467cf9ddf3d084b53d8718c5e2 from
```
                maxItems: 2
```
to
```
                maxItems: 3
```
leads to 'exceeds budget by factor of 1.061681x':
```
ubuntu@ip-172-31-5-54:kubestellar$ kc apply -f config/crd/bases/control.kubestellar.io_bindingpolicies.yaml 
The CustomResourceDefinition "bindingpolicies.control.kubestellar.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[clusterSelectors].items.x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of 1.061681x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```

Changing the same example from
```
                    rule: self.matchLabels.all(key, self.matchLabels[key].startsWith('start-of-value'))
```
to
```
                    rule: self.matchLabels.all(key, self.matchLabels[key].matches('(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?'))
```
leads to 'exceeds budget by factor of more than 100x':
```
ubuntu@ip-172-31-5-54:kubestellar$ kc apply -f config/crd/bases/control.kubestellar.io_bindingpolicies.yaml 
The CustomResourceDefinition "bindingpolicies.control.kubestellar.io" is invalid: 
* spec.validation.openAPIV3Schema.properties[spec].properties[clusterSelectors].items.x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
* spec.validation.openAPIV3Schema.properties[spec].properties[clusterSelectors].items.x-kubernetes-validations[0].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema
* spec.validation.openAPIV3Schema: Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```

Validation using regex seems to be very expensive in the CEL based validations, from [here](https://kubernetes.io/docs/reference/using-api/cel/#resource-constraints):
```
For example: operations that use regular expressions, such as match and find, are estimated using an approximated cost of length(regexString)*length(inputString). The approximated cost reflects the worst case time complexity of Go's RE2 implementation.
```

I tried many regexes, the only one which falls within the cost budget is
```
                    rule: self.matchLabels.all(key, self.matchLabels[key].matches(''))
```
This one has `length(regexString)==0` so it is within the budget, but it is apparently useless.

I think the authors of CEL based validations put very high priority on the performance to protect the API server. As a result, people are complaining about the cost estimations, e.g. https://github.com/kubernetes/kubernetes/issues/120973

#### Summary
If (G) is not viable, then all other options which end up with CEL based validations are all not.

BTW, in response to @MikeSpreitzer 's question in (D). I was not saying `metav1.LabelSelector` is buggy. What I was saying is that: The APIServer makes a decision *not* to strictly validate the occurrences of `metav1.LabelSelector` in a CR object's spec. This decision probably was made because of the same high priority put on performance.

**Comment by MikeSpreitzer** [comment]
@waltforme : thanks for the experiment. But what you tried is (B), not (G). In G I am suggesting to use OpenAPI schema to express the additional constraint, not use CEL.

A full solution has to modify the schema for `matchExpressions` as well as `matchLabels`, because the issue appears in both.

To narrow the validation of label values in `matchLabels` would mean changing the `additionalProperties: string` to use a pattern in place of simply `string`.

**Comment by MikeSpreitzer** [comment]
I am talking about changing https://github.com/kubestellar/kubestellar/blob/v0.21.2/config/crd/bases/control.kubestellar.io_bindingpolicies.yaml#L95-L96 , for example. We could add a `pattern` as a sibling to that `type`.

**Comment by MikeSpreitzer** [comment]
For example, we might add the following.

```yaml
pattern: "(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?"
```

**Comment by MikeSpreitzer** [comment]
Note: since kubebuilder set this project up to apply kustomize to the output of controller-gen, it should be fairly easy to extend the kustomization to fix all the places where a LabelSelector appears.

**Comment by waltforme** [comment]
> For example, we might add the following.
> 
> ```yaml
> pattern: "(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?"
> ```

Spent past two hours trying this.

The issue I ran into was that using this regex (which is exactly the same as the one in the error message in the opening commit) doesn't stop the creation of a bindingpolicy with bad `matchLabels` values.
```
ubuntu@ip-172-31-5-54:kubestellar$ . ~/debug/kubestellar/validation/update_crd.sh 
customresourcedefinition.apiextensions.k8s.io/bindingpolicies.control.kubestellar.io configured
pattern: (([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?
type: string
ubuntu@ip-172-31-5-54:kubestellar$ kc create -f ~/debug/kubestellar/validation/bp_with_bad_matchLabels.yaml 
bindingpolicy.control.kubestellar.io/nvidia-gpu-placement created
ubuntu@ip-172-31-5-54:kubestellar$ kc get bp nvidia-gpu-placement -oyaml | yq .spec.clusterSelectors
- matchLabels:
    name: edgeplatform-test-wec1, edgeplatform-test-wec2
```

Until I dig into Swagger documentation https://swagger.io/docs/specification/data-models/data-types/, and noticed that the example shows `pattern: '^\d{3}-\d{2}-\d{4}$'` and the text says `Without ^…$, pattern works as a partial match`. Then I realized there are subtle diff between the matching in the opening comment and the matching here. So turns out that
```
                        pattern: "^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$"
```
stops the creation of the bad bp:
```
ubuntu@ip-172-31-5-54:kubestellar$ . ~/debug/kubestellar/validation/update_crd.sh 
customresourcedefinition.apiextensions.k8s.io/bindingpolicies.control.kubestellar.io configured
pattern: ^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$
type: string
ubuntu@ip-172-31-5-54:kubestellar$ kc create -f ~/debug/kubestellar/validation/bp_with_bad_matchLabels.yaml 
The BindingPolicy "nvidia-gpu-placement" is invalid: spec.clusterSelectors[0].matchLabels.name: Invalid value: "edgeplatform-test-wec1, edgeplatform-test-wec2": spec.clusterSelectors[0].matchLabels.name in body should match '^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?$'
```
So (G) finally works. Will continue on tomorrow.

**Comment by MikeSpreitzer** [comment]
@waltforme: great news, thanks!

I have opened an issue upstream for this, see https://github.com/kubernetes-sigs/controller-tools/issues/914 .
